src/flop.sol

```diff
@@ -135,6 +139,16 @@ contract Flopper is LibNote {
 
         vat.move(msg.sender, bids[id].guy, bid);
 
+        // on first dent, clear Ash
+        if (bids[id].tic == 0) {
+            uint Ash = VowLike(bids[id].guy).Ash();
+            if (Ash >= bid) {
+                VowLike(bids[id].guy).kiss(bid);
+            } else {
+                VowLike(bids[id].guy).kiss(Ash);
+            }
+        }
+
         bids[id].guy = msg.sender;
         bids[id].lot = lot;
         bids[id].tic = add(uint48(now), ttl);
```

Comment:
This makes modelling this code harder, because a state update happen _after_ a potential call to `Vow.kiss`. So I would need to split `dent` up into two steps, so that I can do `Flop.dent_init`, then `Vow.kiss`, then `Flop.dent_finish` to finish the state update.

Is there any chance that the added clearing of the ash can happen _after_ the state updates to this contract (the updates to `bid[id] ...`? I think it would only require storing `origGuy = bids[id].guy`. Either way, that's likely how I'll do it in the K.

Also, can the internal:

```
            if (Ash >= bid) {
                VowLike(bids[id].guy).kiss(bid);
            } else {
                VowLike(bids[id].guy).kiss(Ash);
            }
```

Be replaced with a single line:

```
VowLike(bids[id].guy).kiss(min(bid, Ash))
```

Reply:
This also, unfortunately, creates a circular dependency between `Flop` and `Vow`.

Reply:
I'm also looking at `VowLike(bids[id].guy)`, which looks quite weird to me. I guess we're relying on the fact that only the `Vow` can `kick` an auction, so when `tic == 0`, then we know that the `guy` is the `Vow`.

But there also are no guards against the `ttl` being set to 0, so you could conceivably have an admin set `TTL` to `0`, then a normal `Flop . dent` would go through and change `bids[id].guy` to be something other than `Vow` (whoever calls `Flop . dent`). AFAICT, this would cause the call to `kiss` to fail, and then the `dent` too fail, so then `dent`ing that auction would be blocked by the fact that the `kiss` can't go through.

Reply:
Here is the structural change I need to make to the K to avoid the circular dependency: https://github.com/makerdao/mkr-mcd-spec/commit/eea7db11f02f07eb300e50e8903edad14eed5aad

And here is the change I make to `dent` to implement this fix: https://github.com/makerdao/mkr-mcd-spec/commit/bdd3d177acbd91c107055f7e5a8df4f5a0eb383e

Reply:
We could also require `ttl > 0` in the `file` method.

Reply:
Another option would be to make the `vow-addr` a `file`able parameter, and use that instead of `bids[id].guy`. This would also eliminate the problem with the `ttl == 0`, because we wouldn't be relying on the first bid being from the `Vow`.

Reply:
If it was file-able, we could just check that the `guy == vow-addr`, instead of comparing with `tic`.

Reply:
Fair point about `ttl`--while hopefully governance would never make the mistake of setting it to zero (this makes auctions `undealable`, so it would break the contract even without this change), and malicious governance can do _much worse_ things (e.g. steal all collateral), it doesn't hurt to foolproof a bit and require a non-zero `ttl` in `file`.

Adding a `file` capability for the Vow's address was considered when fixing a previous issue (`yank` bug) and is worth considering. The worst thing about it is that it creates this tension where there's an overwhelming desire to simplify the code--`require(msg.sender == vow);` in `kick`, get rid of the `gal` argument to `kick`, etc. It also comes with it's own misconfiguration risk. But it's not a bad option, especially since the Vow's address is needed post-`cage` (and right now is obtained by assuming the Vow is the caller of `cage`). Will think about this more.

Re: `min`...it doesn't save many LoC since there's only one usage and makes the code less local, but OTOH I guess everyone knows what a function named `min` does so it doesn't really decrease readability.

Reply:
Looks good to me!